core/src/test/java/org/elasticsearch/common/util/IndexFolderUpgraderTests.java

```diff
@@ -0,0 +1,408 @@
+        try (NodeEnvironment nodeEnv = newNodeEnvironment(nodeSettings)) {
+            final Index index = new Index(randomAsciiOfLength(10), Strings.randomBase64UUID());
+            Settings settings = Settings.builder()
+                .put(nodeSettings)
+                .put(IndexMetaData.SETTING_INDEX_UUID, index.getUUID())
+                .put(IndexMetaData.SETTING_VERSION_CREATED, Version.V_2_0_0)
+                .put(IndexMetaData.SETTING_DATA_PATH, customPath.toAbsolutePath().toString())
+                .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, randomIntBetween(1, 5))
+                .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 0)
+                .build();
+            IndexMetaData indexState = IndexMetaData.builder(index.getName()).settings(settings).build();
+            int numIdxFiles = randomIntBetween(1, 5);
+            int numTranslogFiles = randomIntBetween(1, 5);
+            IndexSettings indexSettings = new IndexSettings(indexState, nodeSettings);
+            writeIndex(nodeEnv, indexSettings, numIdxFiles, numTranslogFiles);
+            IndexFolderUpgrader helper = new IndexFolderUpgrader(settings, nodeEnv);
+            helper.upgrade(indexSettings.getIndex().getName());
+            checkIndex(nodeEnv, indexSettings, numIdxFiles, numTranslogFiles);
+        }
+    }
+
+    /**
+     * tests upgrade on partially upgraded index, when we crash while upgrading
+     */
+    public void testPartialUpgradeCustomDataPath() throws IOException {
+        Path customPath = createTempDir();
+        final Settings nodeSettings = Settings.builder()
+            .put(NodeEnvironment.ADD_NODE_ID_TO_CUSTOM_PATH.getKey(), randomBoolean())
+            .put(Environment.PATH_SHARED_DATA_SETTING.getKey(), customPath.toAbsolutePath().toString()).build();
+        try (NodeEnvironment nodeEnv = newNodeEnvironment(nodeSettings)) {
+            final Index index = new Index(randomAsciiOfLength(10), Strings.randomBase64UUID());
+            Settings settings = Settings.builder()
+                .put(nodeSettings)
+                .put(IndexMetaData.SETTING_INDEX_UUID, index.getUUID())
+                .put(IndexMetaData.SETTING_VERSION_CREATED, Version.V_2_0_0)
+                .put(IndexMetaData.SETTING_DATA_PATH, customPath.toAbsolutePath().toString())
+                .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, randomIntBetween(1, 5))
+                .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 0)
+                .build();
+            IndexMetaData indexState = IndexMetaData.builder(index.getName()).settings(settings).build();
+            int numIdxFiles = randomIntBetween(1, 5);
+            int numTranslogFiles = randomIntBetween(1, 5);
+            IndexSettings indexSettings = new IndexSettings(indexState, nodeSettings);
+            writeIndex(nodeEnv, indexSettings, numIdxFiles, numTranslogFiles);
+            IndexFolderUpgrader helper = new IndexFolderUpgrader(settings, nodeEnv);
```

Comment:
Thanks for the suggestion, in the test we randomly throw IOE in the first pass and check if upgrade on partially upgraded indices behave as expected